Added support for multi-arch i.e. s390x & pp64cle

modassarrana89 opened the pull request:
Added support for multi-arch i.e. s390x & pp64cle
## Description
1. Added multi-arch build support in Makefile 
2. Updated build and push image github workflow to push multi-arch image
3. Added timeout for golanci-lint as required running under qemu

## How Has This Been Tested?
1. Run Github workflow with local branch & able to push multi-arch to local repository
2. Link for succesful github workflow - https://github.com/modassarrana89/model-registry/actions/runs/8659952670
3. Attached screenshot for reference
<img width="1706" alt="Screenshot 2024-04-12 at 16 37 51" src="https://github.com/kubeflow/model-registry/assets/119933819/e7b05547-e163-41f7-9cdd-bfd598e2618b">

tarilabs reviewed (commented):
Thank you @modassarrana89 for this PR !

Some very early comments:
- can you elaborate on why you are suggesting this approach in place of `scripts/build_deploy.sh` which was used at present, please?
- would it be possible to include ARM, since this multi-arch build is being performed?

thank you as this improves on:
- https://github.com/kubeflow/model-registry/issues/54#:~:text=modify%20Docker%20file%2C%20makefile%2C%20etc%20to%20account%20for%20cross%2Dplatform%20build%20(see%20comment%20for%20examples)

/ok-to-test

tarilabs commented on Makefile:
```diff
@@ -219,4 +220,16 @@ image/build:
+.PHONY: image/buildx
+image/buildx: ## Build and push docker image for cross-platform support
+        # copy existing Dockerfile and insert --platform=${BUILDPLATFORM} into Dockerfile.cross, and preserve the original Dockerfile
+	sed -e '1 s/\(^FROM\)/FROM --platform=\$$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$$\{BUILDPLATFORM\}/' ./Dockerfiles/Dockerfile > ./Dockerfiles/Dockerfile.cross
```
```suggestion
	sed -e '1 s/\(^FROM\)/FROM --platform=\$$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$$\{BUILDPLATFORM\}/' ./Dockerfile > ./Dockerfile.cross
```

tarilabs replied:
the above will at least kick-start the image/buildx, but later fails with:

```
make image/buildx
sed -e '1 s/\(^FROM\)/FROM --platform=\$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$\{BUILDPLATFORM\}/' ./Dockerfiles/Dockerfile > ./Dockerfiles/Dockerfile.cross
/bin/sh: ./Dockerfiles/Dockerfile.cross: No such file or directory
make: *** [image/buildx] Error 1
model-registry % sed -e '1 s/\(^FROM\)/FROM --platform=\$$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$$\{BUILDPLATFORM\}/' ./Dockerfiles/Dockerfile > ./Dockerfiles/Dockerfile.cross
zsh: no such file or directory: ./Dockerfiles/Dockerfile.cross
model-registry % code .
model-registry % make image/buildx                                                                                                                                                            
sed -e '1 s/\(^FROM\)/FROM --platform=\$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$\{BUILDPLATFORM\}/' ./Dockerfile > ./Dockerfile.cross
docker buildx create --name project-v3-builder
Failed to initialize: unable to resolve docker endpoint: context "colima": context not found: open /Users/mmortari/.docker/contexts/meta/f24fd3749c1368328e2b149bec149cb6795619f244c5b584e844961215dadd16/meta.json: no such file or directory
make: [image/buildx] Error 1 (ignored)
docker buildx use project-v3-builder
ERROR: failed to find instance "project-v3-builder": open /Users/mmortari/.docker/buildx/instances/project-v3-builder: no such file or directory
make: *** [image/buildx] Error 1
model-registry % export DOCKER_HOST=unix:///Users/mmortari/.local/share/containers/podman/machine/qemu/podman.sock
export TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE="/var/run/docker.sock"
model-registry % make image/buildx                                                                                
sed -e '1 s/\(^FROM\)/FROM --platform=\$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$\{BUILDPLATFORM\}/' ./Dockerfile > ./Dockerfile.cross
docker buildx create --name project-v3-builder
project-v3-builder
docker buildx use project-v3-builder
docker buildx build --push --platform=linux/amd64,linux/s390x,linux/ppc64le --tag kubeflow/model-registry:main -f Dockerfile.cross .
[+] Building 173.8s (21/35)                                                                                                                                                                                                                docker-container:project-v3-builder
 => [internal] booting buildkit                                                                                                                                                                                                                                           2.0s
 => => pulling image moby/buildkit:buildx-stable-1                                                                                                                                                                                                                        1.5s
 => => creating container buildx_buildkit_project-v3-builder0                                                                                                                                                                                                             0.5s
 => [internal] load build definition from Dockerfile.cross                                                                                                                                                                                                                0.0s
 => => transferring dockerfile: 1.92kB                                                                                                                                                                                                                                    0.0s
 => [linux/s390x internal] load metadata for registry.access.redhat.com/ubi8/ubi-minimal:8.8                                                                                                                                                                              1.9s
 => [linux/amd64 internal] load metadata for registry.access.redhat.com/ubi8/ubi-minimal:8.8                                                                                                                                                                              1.9s
 => [linux/ppc64le internal] load metadata for registry.access.redhat.com/ubi8/ubi-minimal:8.8                                                                                                                                                                            1.9s
 => [linux/arm64 internal] load metadata for registry.access.redhat.com/ubi8/go-toolset:1.19                                                                                                                                                                              2.9s
 => [internal] load .dockerignore                                                                                                                                                                                                                                         0.0s
 => => transferring context: 52B                                                                                                                                                                                                                                          0.0s
 => [linux/s390x stage-1 1/3] FROM registry.access.redhat.com/ubi8/ubi-minimal:8.8@sha256:b93deceb59a58588d5b16429fc47f98920f84740a1f2ed6454e33275f0701b59                                                                                                               31.6s
 => => resolve registry.access.redhat.com/ubi8/ubi-minimal:8.8@sha256:b93deceb59a58588d5b16429fc47f98920f84740a1f2ed6454e33275f0701b59                                                                                                                                    0.0s
 => => sha256:8737e1657f05079caa8c726fb829e674f98ab5a3dd612e1bbd5a3b46b095e360 37.55MB / 37.55MB                                                                                                                                                                         31.1s
 => => extracting sha256:8737e1657f05079caa8c726fb829e674f98ab5a3dd612e1bbd5a3b46b095e360                                                                                                                                                                                 0.5s
 => [internal] load build context                                                                                                                                                                                                                                         0.3s
 => => transferring context: 6.77MB                                                                                                                                                                                                                                       0.3s
 => [linux/arm64 builder  1/21] FROM registry.access.redhat.com/ubi8/go-toolset:1.19@sha256:dd16d34f7e4fae45a620ba8fdde378d0853314e4e0f4998658fd81628b36f0d6                                                                                                             92.4s
 => => resolve registry.access.redhat.com/ubi8/go-toolset:1.19@sha256:dd16d34f7e4fae45a620ba8fdde378d0853314e4e0f4998658fd81628b36f0d6                                                                                                                                    0.0s
 => => sha256:b953f5a6908d8aa51e2200fe3aa4b40102eb758be1d56e7dd3bec2a9cb85662e 162.55MB / 162.55MB                                                                                                                                                                       90.2s
 => => sha256:b8104c5f40fdcf0549bf9ad8b9ea4e5d3fabcc446f920cb976986f255e884a42 143.61MB / 143.61MB                                                                                                                                                                       70.0s
 => => sha256:8550ec08a06cf30baf540ef95dcc7334958355ef556b38eca4fd67cd68de5af1 18.17MB / 18.17MB                                                                                                                                                                         13.0s
 => => sha256:e1bb0572465a9e03d7af5024abb36d7227b5bf133c448b54656d908982127874 76.88MB / 76.88MB                                                                                                                                                                         43.1s
 => => extracting sha256:e1bb0572465a9e03d7af5024abb36d7227b5bf133c448b54656d908982127874                                                                                                                                                                                 1.2s
 => => extracting sha256:8550ec08a06cf30baf540ef95dcc7334958355ef556b38eca4fd67cd68de5af1                                                                                                                                                                                 0.3s
 => => extracting sha256:b8104c5f40fdcf0549bf9ad8b9ea4e5d3fabcc446f920cb976986f255e884a42                                                                                                                                                                                 2.2s
 => => extracting sha256:b953f5a6908d8aa51e2200fe3aa4b40102eb758be1d56e7dd3bec2a9cb85662e                                                                                                                                                                                 2.2s
 => [linux/ppc64le stage-1 1/3] FROM registry.access.redhat.com/ubi8/ubi-minimal:8.8@sha256:b93deceb59a58588d5b16429fc47f98920f84740a1f2ed6454e33275f0701b59                                                                                                             17.1s
 => => resolve registry.access.redhat.com/ubi8/ubi-minimal:8.8@sha256:b93deceb59a58588d5b16429fc47f98920f84740a1f2ed6454e33275f0701b59                                                                                                                                    0.0s
 => => sha256:655f24e7656ea1c1a2711179ea0c6d6f7dac72ba7d2d3fd34c36d062f00f74d8 44.03MB / 44.03MB                                                                                                                                                                         16.5s
 => => extracting sha256:655f24e7656ea1c1a2711179ea0c6d6f7dac72ba7d2d3fd34c36d062f00f74d8                                                                                                                                                                                 0.6s
 => [linux/amd64 stage-1 1/3] FROM registry.access.redhat.com/ubi8/ubi-minimal:8.8@sha256:b93deceb59a58588d5b16429fc47f98920f84740a1f2ed6454e33275f0701b59                                                                                                               31.4s
 => => resolve registry.access.redhat.com/ubi8/ubi-minimal:8.8@sha256:b93deceb59a58588d5b16429fc47f98920f84740a1f2ed6454e33275f0701b59                                                                                                                                    0.0s
 => => sha256:dc35b837139a95d1b9f7f7b0435a024a74ab972416bdc248f3f608c9f917a753 39.31MB / 39.31MB                                                                                                                                                                         30.8s
 => => extracting sha256:dc35b837139a95d1b9f7f7b0435a024a74ab972416bdc248f3f608c9f917a753                                                                                                                                                                                 0.5s
 => [linux/arm64 builder  2/21] WORKDIR /workspace                                                                                                                                                                                                                        0.3s
 => [linux/arm64 builder  3/21] COPY [go.mod, go.sum, ./]                                                                                                                                                                                                                 0.0s
 => [linux/arm64 builder  4/21] RUN go mod download                                                                                                                                                                                                                      17.6s
 => [linux/arm64 builder  5/21] RUN yum remove -y nodejs npm                                                                                                                                                                                                              0.8s
 => [linux/arm64 builder  6/21] RUN yum module -y reset nodejs                                                                                                                                                                                                            6.2s
 => [linux/arm64 builder  7/21] RUN yum module -y enable nodejs:18                                                                                                                                                                                                        0.4s
 => [linux/arm64 builder  8/21] RUN yum install -y nodejs npm java-11                                                                                                                                                                                                    50.6s
 => [linux/arm64 builder  9/21] COPY [Makefile, main.go, .openapi-generator-ignore, openapitools.json, ./]                                                                                                                                                                0.1s
 => ERROR [linux/arm64 builder 10/21] RUN set -ex    ; ARCH=$(uname -m) ; echo $ARCH     ; if [ "$ARCH" == "s390x" ] ; then ARCH="s390_64" ; elif [ "$ARCH" == "ppc64le" ] ; then ARCH="ppcle_64" ; fi     ; wget -q https://github.com/protocolbuffers/protobuf/release  0.4s
------
 > [linux/arm64 builder 10/21] RUN set -ex    ; ARCH=$(uname -m) ; echo $ARCH     ; if [ "$ARCH" == "s390x" ] ; then ARCH="s390_64" ; elif [ "$ARCH" == "ppc64le" ] ; then ARCH="ppcle_64" ; fi     ; wget -q https://github.com/protocolbuffers/protobuf/releases/download/v24.3/protoc-24.3-linux-$ARCH.zip -O protoc.zip     ; unzip -q protoc.zip     ; bin/protoc --version     ; rm protoc.zip     ;:
0.037 ++ uname -m
0.039 + ARCH=aarch64
0.039 + echo aarch64
0.039 + '[' aarch64 == s390x ']'
0.039 + '[' aarch64 == ppc64le ']'
0.039 + wget -q https://github.com/protocolbuffers/protobuf/releases/download/v24.3/protoc-24.3-linux-aarch64.zip -O protoc.zip
0.039 aarch64
------
Dockerfile.cross:24
--------------------
  23 |     # Download protoc compiler v24.3
  24 | >>> RUN set -ex\
  25 | >>>     ; ARCH=$(uname -m) ; echo $ARCH \
  26 | >>>     ; if [ "$ARCH" == "s390x" ] ; then ARCH="s390_64" ; elif [ "$ARCH" == "ppc64le" ] ; then ARCH="ppcle_64" ; fi \
  27 | >>>     ; wget -q https://github.com/protocolbuffers/protobuf/releases/download/v24.3/protoc-24.3-linux-$ARCH.zip -O protoc.zip \
  28 | >>>     ; unzip -q protoc.zip \
  29 | >>>     ; bin/protoc --version \
  30 | >>>     ; rm protoc.zip \
  31 | >>>     ;
  32 |     
--------------------
ERROR: failed to solve: process "/bin/sh -c set -ex    ; ARCH=$(uname -m) ; echo $ARCH     ; if [ \"$ARCH\" == \"s390x\" ] ; then ARCH=\"s390_64\" ; elif [ \"$ARCH\" == \"ppc64le\" ] ; then ARCH=\"ppcle_64\" ; fi     ; wget -q https://github.com/protocolbuffers/protobuf/releases/download/v24.3/protoc-24.3-linux-$ARCH.zip -O protoc.zip     ; unzip -q protoc.zip     ; bin/protoc --version     ; rm protoc.zip     ;" did not complete successfully: exit code: 8
make: [image/buildx] Error 1 (ignored)
docker buildx rm project-v3-builder
project-v3-builder removed
rm Dockerfile.cross
```

modassarrana89 replied:
I have tested with s390x , pp64cle & amd64. its failing because above testing has been done for arm64. arm64 support is not there in Makefile.

tarilabs replied:
I'm not understanding how it can work without https://github.com/kubeflow/model-registry/pull/62/files#r1562588431

can you elaborate?

modassarrana89 replied:
There is small confusion from my end. I have put my changes on 1 week old code . I will rebase my change on to the latest code & raise again. As part of that your abnove comments will be addressed

tarilabs replied:
No problem, thank you!

modassarrana89 replied:
Handle it in the latest changes . @tarilabs please review

tarilabs commented on Dockerfile:
```diff
@@ -19,6 +19,20 @@ RUN yum install -y nodejs npm java-11
 COPY ["Makefile", "main.go", ".openapi-generator-ignore", "openapitools.json", "./"]
 
+
+# Download protoc compiler v24.3
+RUN set -ex\
+    ; ARCH=$(uname -m) ; echo $ARCH \
+    ; if [ "$ARCH" == "s390x" ] ; then ARCH="s390_64" ; elif [ "$ARCH" == "ppc64le" ] ; then ARCH="ppcle_64" ; fi \
+    ; wget -q https://github.com/protocolbuffers/protobuf/releases/download/v24.3/protoc-24.3-linux-$ARCH.zip -O protoc.zip \
+    ; unzip -q protoc.zip \
+    ; bin/protoc --version \
+    ; rm protoc.zip \
+    ;
```
The download of protoc is already performed as part of `make deps` which includes `make bin/protoc`, in line below 34.

This was improved with #52 

Kindly consider extending `scripts/install_protoc.sh` for your platforms, so we don't need this extra Docker RUN ?

modassarrana89 replied:
i didn't know about the protoc change . Once it is approved , i will update my changes accordingly for protoc

tarilabs commented on Dockerfile:
```diff
@@ -19,6 +19,20 @@ RUN yum install -y nodejs npm java-11
+    ;
+
+# Download tools
+RUN make deps
```
since this `RUN make deps` happens-before copying `COPY scripts/ scripts/` (line 27/41) it will make makefile fail to execute:

https://github.com/kubeflow/model-registry/blob/5f7fac0f1d98dc3daa0027801dc365ac7e26814c/Makefile#L115-L116

and it's actually a duplicate of line 33/47

modassarrana89 replied:
I will update install_protoc with other architecture.

modassarrana89 replied:
There is small confusion from my end. I have put my changes on 1 week old code . I will rebase my change on to the latest code & raise again. As part of that your abnove comments will be addressed

modassarrana89 replied:
Updated protoc.sh scripts for all architecture. @tarilabs  Please review.

tarilabs reviewed (commented):
> Once it is approved , i will update my changes accordingly for protoc

Hi, I cannot approve until the setup can be replicated and [sign-off commits](https://wiki.linuxfoundation.org/dco) to comply with DCO as required by KF community.

It seems to me at least the following modifications are needed:

```diff
diff --git a/Dockerfile b/Dockerfile
index 5980940..d777709 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -19,20 +19,6 @@ RUN yum install -y nodejs npm java-11
 # Copy the go source
 COPY ["Makefile", "main.go", ".openapi-generator-ignore", "openapitools.json", "./"]
 
-
-# Download protoc compiler v24.3
-RUN set -ex\
-    ; ARCH=$(uname -m) ; echo $ARCH \
-    ; if [ "$ARCH" == "s390x" ] ; then ARCH="s390_64" ; elif [ "$ARCH" == "ppc64le" ] ; then ARCH="ppcle_64" ; fi \
-    ; wget -q https://github.com/protocolbuffers/protobuf/releases/download/v24.3/protoc-24.3-linux-$ARCH.zip -O protoc.zip \
-    ; unzip -q protoc.zip \
-    ; bin/protoc --version \
-    ; rm protoc.zip \
-    ;
-
-# Download tools
-RUN make deps
-
 # Copy rest of the source
 COPY .git/ .git/
 COPY cmd/ cmd/
diff --git a/Makefile b/Makefile
index a227ee8..bba3fc7 100644
--- a/Makefile
+++ b/Makefile
@@ -225,7 +225,7 @@ PLATFORMS ?= linux/amd64,linux/s390x,linux/ppc64le
 .PHONY: image/buildx
 image/buildx: ## Build and push docker image for cross-platform support
         # copy existing Dockerfile and insert --platform=${BUILDPLATFORM} into Dockerfile.cross, and preserve the original Dockerfile
-       sed -e '1 s/\(^FROM\)/FROM --platform=\$$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$$\{BUILDPLATFORM\}/' ./Dockerfiles/Dockerfile > ./Dockerfiles/Dockerfile.cross
+       sed -e '1 s/\(^FROM\)/FROM --platform=\$$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$$\{BUILDPLATFORM\}/' ./Dockerfile > ./Dockerfile.cross
        - $(DOCKER) buildx create --name project-v3-builder
        $(DOCKER) buildx use project-v3-builder
        - $(DOCKER) buildx build --push --platform=$(PLATFORMS) --tag ${IMG}:$(IMG_VERSION) -f Dockerfile.cross .
diff --git a/scripts/install_protoc.sh b/scripts/install_protoc.sh
index b7b165e..e346c14 100755
--- a/scripts/install_protoc.sh
+++ b/scripts/install_protoc.sh
@@ -12,6 +12,10 @@ fi
 ARCH="x86_64"
 if [[ "$(uname -m)" == "arm"* ]]; then
   ARCH="aarch_64"
+elif [[ "$(uname -m)" == "s390x" ]]; then
+  ARCH="s390_64"
+elif [[ "$(uname -m)" == "ppc64le" ]]; then
+  ARCH="ppcle_64"
 fi
 
 mkdir -p ${SCRIPT_DIR}/../bin

```

as in:
1. keep using `scripts/install_protoc.sh` for downloading protoc as part of makefile, not re-download as part of dockerfile
2. not sure why this original PR changes makefile to use dockerfile in subdirectory `./Dockerfiles/` 🤔 

Also, I believe it would be ideal for makefile `image/buildx` to not push the images, wdyt?

We appreciate your PR a lot! We just need to account for these comments and kindly sign-off commits so to make it easier to collaborate.

modassarrana89 commented:
@tarilabs I have added architecture changes in install_protoc.sh , removed protoc part from dockerfile , updated Makefile as well. Addressed all your comments . Please review once.

tarilabs commented:
> @tarilabs I have added architecture changes in install_protoc.sh , removed protoc part from dockerfile , updated Makefile as well. Addressed all your comments . Please review once.

thank you for the updates @modassarrana89 ! As mentioned last [Apr 12](https://github.com/kubeflow/model-registry/pull/62#pullrequestreview-1996634293), and last [Apr 14](https://github.com/kubeflow/model-registry/pull/62#pullrequestreview-1999732920), ensure you [signing-off your commits](https://wiki.linuxfoundation.org/dco) to comply with DCO, KF community has decided it's being used going forward. This is required.

tarilabs commented:
> i am getting issues with your fix . Do you mind , if i cancelled this PR & actually create a new one with my changes & proper signoff ( as required ). Please suggest

not sure which of "my" fix you are referring to, but feel free to close this PR and reopen a new one, it's no problem! thank you

modassarrana89 commented:
@tarilabs Closing this PR , as another one( https://github.com/kubeflow/model-registry/pull/66 ) has been raised with the required changes
